internals/overlord/logstate/gatherer.go

```diff
@@ -0,0 +1,255 @@
+		}
+
+		buffer, bufferExists := buffers[service.Name]
+		if !bufferExists {
+			// We don't yet have a reference to the service's ring buffer
+			// Need to wait until serviceStarted
+			continue
+		}
+
+		// Create new puller
+		newPullers[service.Name] = newLogPuller(g.entryCh, buffer)
+	}
+
+	// Old pullers for now-removed services need to be shut down.
+	for _, puller := range g.pullers {
+		_ = puller.tomb.Killf("plan changed")
+	}
+	g.pullers = newPullers
+}
+
+func (g *logGatherer) serviceStarted(service *plan.Service, buffer *servicelog.RingBuffer) {
+	g.pullersLock.Lock()
+	defer g.pullersLock.Unlock()
+
+	if puller, ok := g.pullers[service.Name]; ok {
+		// This should never happen, but just in case, shut down the old puller.
+		_ = puller.tomb.Killf("service started")
+	}
+	g.pullers[service.Name] = newLogPuller(g.entryCh, buffer)
+}
+
+func (g *logGatherer) loop() error {
+	ticker := time.NewTicker(g.tickPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-g.tomb.Dying():
+			return nil
+
+		case <-ticker.C:
+			// Timeout - flush
+			err := g.client.Flush(g.clientCtx)
+			if err != nil {
+				return fmt.Errorf("sending logs to target %q: %v", g.targetName, err)
+			}
+
+		case entry := <-g.entryCh:
+			err := g.client.Write(g.clientCtx, entry)
+			if err != nil {
+				return fmt.Errorf("writing logs to target %q: %v", g.targetName, err)
+			}
+		}
+	}
+}
+
+// stop closes the cancel channel, thereby terminating the main loop.
+// This method will block until gatherer teardown is complete.
+func (g *logGatherer) stop() {
+	// Start a timeout
+	time.AfterFunc(teardownTimeout, g.clientCancel)
+	// Kill the main loop
+	_ = g.tomb.Killf("gatherer closed")
+
+	// Shut down pullers
+	g.pullersLock.Lock()
+	defer g.pullersLock.Unlock()
+	for _, puller := range g.pullers {
+		_ = puller.tomb.Killf("gatherer closed")
+	}
+
+	// Wait for the main loop to finish - then we can do a final flush of the client
+	_ = g.tomb.Wait()
+	err := g.client.Flush(g.clientCtx)
+	if err != nil {
+		logger.Noticef("Cannot send logs to target %q: %v", g.targetName, err)
+	}
+}
+
+// logPuller handles pulling logs from a single iterator and sending to the
+// main control loop.
+type logPuller struct {
+	iterator servicelog.Iterator
+	tomb     *tomb.Tomb
+	entryCh  chan servicelog.Entry
+}
+
+func newLogPuller(entryCh chan servicelog.Entry, buffer *servicelog.RingBuffer) *logPuller {
+	p := &logPuller{
+		iterator: buffer.TailIterator(),
+		tomb:     &tomb.Tomb{},
+		entryCh:  entryCh,
+	}
+	p.tomb.Go(p.loop)
+	return p
+}
+
+func (p *logPuller) loop() error {
+	parser := servicelog.NewParser(p.iterator, parserSize)
+
+iteratorLoop:
+	for p.iterator.Next(p.tomb.Dying()) {
+		for parser.Next() {
+			if err := parser.Err(); err != nil {
+				return err
+			}
+			select {
+			case p.entryCh <- parser.Entry():
+			case <-p.tomb.Dying():
+				break iteratorLoop
+			}
+		}
+	}
+
+	// We've been killed - close the iterator
+	return p.iterator.Close()
```

Comment:
I guess we could `defer p.iterator.Close()` at the top of this function, and then the `break iteratorLoop` could become just `return` and we wouldn't need the label.